core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -1288,6 +1288,30 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
     mapSizesByExecutorId.iter
   }
 
+  def getMapOutputLocationWithRefresh(
```

Comment:
Maybe return `Option[BlockManagerId]`?: 
```
  def getMapOutputLocationWithRefresh(
                                       shuffleId: Int,
                                       mapId: Long,
                                       prevLocation: BlockManagerId): Option[BlockManagerId] = {
    // Try to get the cached location first in case other concurrent tasks
    // fetched the fresh location already
    getMapOutputLocation(shuffleId, mapId) match {
      case Some(location) =>
        if (location == prevLocation) {
          unregisterShuffle(shuffleId)
          getMapOutputLocation(shuffleId, mapId)
        } else {
          None
        }
      case _ =>
        None
    }
  }
```

Reply:
We can do the following with `Option`: 
```
                val currentAddressOpt = mapOutputTrackerWorker
                  .getMapOutputLocationWithRefresh(shuffleId, mapId, address)
                currentAddressOpt match {
                  case Some(currentAddress) =>
                    if (currentAddress != address) {
                      logInfo(s"Map status location for block $blockId changed from $address " +
                        s"to $currentAddress")
                      remainingBlocks -= blockId
                      deferredBlocks.getOrElseUpdate(currentAddress, new ArrayBuffer[String]())
                        .append(blockId)
                      enqueueDeferredFetchRequestIfNecessary()
                    } else {
                      results.put(FailureFetchResult(block, infoMap(blockId)._2, address, e))
                    }
                  case None =>
                    results.put(FailureFetchResult(block, infoMap(blockId)._2, address, e))
                }
```
It is also consistent with other function signatures with `getMapOutputLocation`.